File: Codex/Rust/2025/Day12_Solution_2025.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("Day12_Input.txt").expect("failed to read Day12_Input.txt");
    let mut shape_areas = Vec::new();
    let mut lines = input.lines().peekable();

    while let Some(line) = lines.peek().copied() {
        let trimmed = line.trim();
        if trimmed.is_empty() {
            lines.next();
            continue;
        }

        if trimmed.contains('x') {
            break;
        }

        if trimmed.ends_with(':') {
            lines.next();
            let mut area = 0_u64;
            while let Some(shape_line) = lines.peek().copied() {
                let shape_trimmed = shape_line.trim();
                if shape_trimmed.is_empty() {
                    lines.next();
                    break;
                }
                if shape_trimmed.ends_with(':') || shape_trimmed.contains('x') {
                    break;
                }
                area += shape_trimmed.bytes().filter(|&b| b == b'#').count() as u64;
                lines.next();
            }
            shape_areas.push(area);
            continue;
        }

        panic!("unexpected input format");
    }

    let mut answer = 0_u64;

    for line in lines {
        let trimmed = line.trim();
        if trimmed.is_empty() {
            continue;
        }

        let (dims, counts_text) = trimmed.split_once(':').expect("invalid region line");
        let (width_text, height_text) = dims.split_once('x').expect("invalid dimensions");
        let width: u64 = width_text.parse().expect("invalid width");
        let height: u64 = height_text.parse().expect("invalid height");
        let counts: Vec<u64> = counts_text
            .split_whitespace()
            .map(|value| value.parse().expect("invalid count"))
            .collect();

        let needed_area: u64 = shape_areas
            .iter()
            .zip(counts.iter())
            .map(|(area, count)| area * count)
            .sum();

        if needed_area <= width * height {
            answer += 1;
        }
    }

    println!("{answer}");
}

File: Codex/Python/2025/Day12_Solution_2025.py
import time

start = time.perf_counter()

def main() -> None:
    with open("Day12_Input.txt", "r", encoding="utf-8") as file:
        lines = [line.rstrip("\n") for line in file]

    shape_areas: list[int] = []
    index = 0

    while index < len(lines):
        line = lines[index].strip()
        if not line:
            index += 1
            continue
        if "x" in line and ":" in line:
            break

        if line.endswith(":"):
            area = 0
            index += 1
            while index < len(lines) and lines[index].strip():
                area += lines[index].count("#")
                index += 1
            shape_areas.append(area)
        else:
            index += 1

    fit_count = 0

    while index < len(lines):
        line = lines[index].strip()
        index += 1
        if not line:
            continue

        size_text, counts_text = line.split(": ")
        width_text, height_text = size_text.split("x")
        width = int(width_text)
        height = int(height_text)
        counts = [int(value) for value in counts_text.split()]

        required_area = sum(count * area for count, area in zip(counts, shape_areas))
        if required_area <= width * height:
            fit_count += 1

    print(fit_count)


if __name__ == "__main__":
    start = time.perf_counter()
    main()
    end = time.perf_counter()
    print(f"Runtime: {end - start:.6f} seconds")

end = time.perf_counter()
print(f"Runtime: {end - start:.6f} seconds")